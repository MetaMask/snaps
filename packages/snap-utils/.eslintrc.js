module.exports = {
  extends: ['../../.eslintrc.js'],

  ignorePatterns: ['src/json-schemas'],
};
