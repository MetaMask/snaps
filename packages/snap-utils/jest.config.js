const deepmerge = require('deepmerge');
const baseConfig = require('../../jest.config.base');

module.exports = deepmerge(baseConfig, {
  coveragePathIgnorePatterns: [
    './src/index.ts',
    './src/index.browser.ts',
    './src/test-utils',
    './src/json-schemas',
    // Jest currently doesn't collect coverage for child processes.
    // https://github.com/facebook/jest/issues/5274
    './src/eval-worker.ts',
  ],
  coverageThreshold: {
    global: {
      branches: 93.21,
      functions: 97.46,
      lines: 98.07,
      statements: 98.07,
    },
  },
  testTimeout: 2500,
});
