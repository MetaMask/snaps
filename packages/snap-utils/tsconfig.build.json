{
  "extends": "../../tsconfig.packages.build.json",
  "compilerOptions": {
    "baseUrl": "./",
    "outDir": "./dist",
    "rootDir": "./src",
    "tsBuildInfoFile": "../../.cache/snap-utils.tsbuildinfo"
  },
  "include": ["./src"],
  "exclude": [
    "**/*.test.ts",
    "./src/**/test-utils",
    "./src/**/__mocks__",
    "./src/**/__snapshots__"
  ]
}
