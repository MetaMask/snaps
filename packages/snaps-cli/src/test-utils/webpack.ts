import { logError } from '@metamask/snaps-utils';
import { DEFAULT_SNAP_BUNDLE } from '@metamask/snaps-utils/test-utils';
import type { IFs } from 'memfs';
import { createFsFromVolume, Volume } from 'memfs';
import { promisify } from 'util';
import type { Configuration } from 'webpack';
import webpack from 'webpack';

export type CompileOptions = {
  code?: string;
  fileSystem?: IFs;
  config?: Configuration;
  watchMode?: boolean;
};

/**
 * Compile a snap bundle using webpack.
 *
 * @param options - Options to configure the compilation.
 * @param options.code - The snap bundle code.
 * @param options.fileSystem - The file system to use. Defaults to an in-memory
 * file system.
 * @param options.config - The webpack configuration to use. Defaults to a
 * minimal configuration that compiles the provided code.
 * @param options.watchMode - Whether to compile in watch mode. Defaults to
 * false.
 * @returns The compiled code.
 */
export async function getCompiler({
  code = DEFAULT_SNAP_BUNDLE,
  fileSystem = createFsFromVolume(new Volume()),
  config,
  watchMode = false,
}: CompileOptions = {}) {
  await fileSystem.promises.writeFile('/input.js', code);

  const compiler = webpack({
    entry: '/input.js',
    mode: 'none',
    output: {
      path: '/',
      filename: 'output.js',
    },
    ...config,
  });

  compiler.watchMode = watchMode;
  compiler.inputFileSystem = fileSystem;
  compiler.outputFileSystem = fileSystem;

  return compiler;
}

/**
 * Compile a snap bundle using webpack.
 *
 * @param options - Options to configure the compilation.
 * @param options.code - The snap bundle code.
 * @param options.fileSystem - The file system to use. Defaults to an in-memory
 * file system.
 * @param options.config - The webpack configuration to use. Defaults to a
 * minimal configuration that compiles the provided code.
 * @returns The compiled code.
 */
export async function compile(
  options: CompileOptions,
): Promise<{ code: string }> {
  const compiler = await getCompiler(options);

  return new Promise((resolve, reject) => {
    compiler.run(async (error, stats) => {
      if (error || !stats) {
        reject(new Error('Failed to compile.'));
        return;
      }

      if (stats.hasErrors()) {
        logError(stats.toString('errors-only'));
        reject(new Error('Failed to compile.'));
        return;
      }

      const output = (await promisify(compiler.outputFileSystem.readFile)(
        '/output.js',
      )) as string;

      resolve({
        code: output,
      });
    });
  });
}

/**
 * Fakes the fallback modules path to ensure that snapshots have the same path on different systems.
 *
 * @param config - The webpack configuration.
 * @returns The webpack configuration with the fallback paths normalized.
 */
export function fakePolyfillPaths(config: Configuration): Configuration {
  if (!config.resolve?.fallback) {
    return config;
  }

  const {
    resolve: { fallback, ...resolveRest },
    ...rest
  } = config;

  const normalizedFallbacks = Object.keys(fallback).reduce((acc, index) => {
    if (typeof (fallback as Record<string, string>)[index] === 'string') {
      return { [index]: '/foo/bar/index.js', ...acc };
    }

    return { [index]: (fallback as Record<string, string>)[index], ...acc };
  }, {});

  return {
    ...rest,
    resolve: { ...resolveRest, fallback: normalizedFallbacks },
  };
}
