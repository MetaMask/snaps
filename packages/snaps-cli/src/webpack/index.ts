export * from './compiler';
export * from './config';
export * from './plugins';
