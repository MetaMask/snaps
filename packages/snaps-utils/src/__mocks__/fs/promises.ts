import { fs } from 'memfs';

export = fs.promises;
