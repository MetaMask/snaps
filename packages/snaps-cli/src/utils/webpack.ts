import { logError, logInfo } from '@metamask/snaps-utils';
import SnapsWebpackPlugin from '@metamask/snaps-webpack-plugin';
import TerserPlugin from 'terser-webpack-plugin';
import { Compiler, Configuration, webpack } from 'webpack';

import { ProcessedWebpackConfig } from '../config';

export type WebpackOptions = {
  /**
   * Whether to watch for changes.
   */
  watch?: boolean;

  /**
   * Whether to evaluate the bundle. If this is set, it will override the
   * `evaluate` option in the config object.
   */
  evaluate?: boolean;
};

export type WatchPluginOptions = {
  /**
   * The extra files to watch.
   */
  files?: string[];
};

export class WatchPlugin {
  #options: WatchPluginOptions;

  constructor(options: WatchPluginOptions) {
    this.#options = options;
  }

  apply(compiler: Compiler) {
    compiler.hooks.afterCompile.tapPromise(
      'WatchPlugin',
      async ({ fileDependencies }) => {
        this.#options.files?.forEach(
          fileDependencies.add.bind(fileDependencies),
        );
      },
    );
  }
}

/**
 * Get the Webpack devtool configuration based on the given snap config.
 *
 * - If `sourceMap` is `inline`, return `inline-source-map`.
 * - If `sourceMap` is `true`, return `source-map`.
 * - Otherwise, return `false`.
 *
 * @param config - The processed snap Webpack config.
 * @param config.sourceMap - Whether to generate source maps.
 * @returns The Webpack devtool configuration.
 */
function getDevTool({
  sourceMap,
}: ProcessedWebpackConfig): Configuration['devtool'] {
  if (sourceMap === 'inline') {
    return 'inline-source-map';
  }

  if (sourceMap === true) {
    return 'source-map';
  }

  return false;
}

/**
 * Get the default Webpack configuration. This is the configuration that will
 * be used if the user doesn't provide a custom Webpack configuration. The
 * configuration is based on the snap config.
 *
 * The default configuration includes:
 *
 * - `SWC` to transpile TypeScript and JavaScript files.
 * - `TerserPlugin` to minify the bundle.
 * - `SnapsWebpackPlugin` to validate the bundle and update the manifest.
 *
 * It can be customized through the `customizeWebpackConfig` function in the
 * snap config, but in most cases, you shouldn't need to do that.
 *
 * @param config - The processed snap Webpack config.
 * @param options - The Webpack options.
 * @returns The default Webpack configuration.
 */
export function getDefaultConfiguration(
  config: ProcessedWebpackConfig,
  options: WebpackOptions = {},
): Configuration {
  return {
    /**
     * `web` is the default target for Webpack, so we don't need to set it, but
     * we do so here for clarity. This ensures that Webpack will build a bundle
     * that can be loaded in a browser.
     *
     * @see https://webpack.js.org/configuration/target/
     */
    target: 'web',

    /**
     * The mode is set to `production` by default, so that Webpack will minify
     * and optimize the bundle.
     *
     * @see https://webpack.js.org/configuration/mode/
     */
    mode: 'production',

    /**
     * The entry point is set to the `entry` value from the config object.
     *
     * @see https://webpack.js.org/configuration/entry-context/
     */
    entry: config.entry,

    /**
     * The devtool option controls how source maps are generated. We set it to
     * the `sourceMap` value from the config object.
     *
     * @see https://webpack.js.org/configuration/devtool/
     */
    devtool: getDevTool(config),

    /**
     * The stats option controls how much information is printed to the console
     * when Webpack is running. We set it to `none` so that we can control the
     * output ourselves.
     *
     * @see https://webpack.js.org/configuration/stats/
     */
    stats: 'none',

    /**
     * The output options.
     *
     * @see https://webpack.js.org/configuration/output/
     */
    output: {
      /**
       * This indicates whether Webpack should clear the output directory
       * before building. We set it to the `clean` value from the config
       * object.
       *
       * @see https://webpack.js.org/configuration/output/#outputclean
       */
      clean: config.output.clean,

      /**
       * The filename of the bundle. We set it to the `filename` value from
       * the config object.
       *
       * @see https://webpack.js.org/configuration/output/#outputfilename
       */
      filename: config.output.filename,

      /**
       * The path to the output directory. We set it to the `path` value from
       * the config object.
       *
       * @see https://webpack.js.org/configuration/output/#outputpath
       */
      path: config.output.path,

      /**
       * The public path of the bundle. We set it to `/` by default, so that
       * the bundle can be loaded from the root of the server.
       *
       * @see https://webpack.js.org/configuration/output/#outputpublicpath
       */
      publicPath: '/',

      /**
       * The library configuration. This tells Webpack how to export the
       * bundle.
       *
       * @see https://webpack.js.org/configuration/output/#outputlibrary
       */
      library: {
        /**
         * This tells Webpack to export the bundle as a CommonJS module. This
         * is necessary for MetaMask Snaps.
         *
         * @see https://webpack.js.org/configuration/output/#outputlibrarytarget
         */
        type: 'commonjs',
      },
    },

    /**
     * The module configuration. This is where we tell Webpack how to handle
     * different types of files.
     *
     * @see https://webpack.js.org/configuration/module/
     */
    module: {
      rules: [
        {
          /**
           * We use the `swc-loader` to transpile TypeScript and JavaScript
           * files. This is a Webpack loader that uses the `SWC` compiler,
           * which is a much faster alternative to Babel and TypeScript's own
           * compiler.
           *
           * @see https://swc.rs/docs/usage/swc-loader
           */
          test: /\.tsx?$/u,
          exclude: /node_modules/u,
          use: {
            loader: 'swc-loader',

            /**
             * The options for the `swc-loader`. These can be overridden in the
             * `.swcrc` file.
             *
             * @see https://swc.rs/docs/configuration/swcrc
             */
            options: {
              /**
               * This tells SWC to generate source maps. We set it to the
               * `sourceMap` value from the config object.
               *
               * This must be enabled if source maps are enabled in the config.
               */
              sourceMaps: Boolean(getDevTool(config)),

              jsc: {
                /**
                 * MetaMask targets ES2020, so we set the target to ES2020. If
                 * you need to support older browsers, you can set the target
                 * to something lower to include the necessary polyfills.
                 *
                 * @see https://swc.rs/docs/configuration/compilation#jsctarget
                 */
                target: 'es2020',

                parser: {
                  /**
                   * This tells the parser to parse TypeScript files. If you
                   * don't need to support TypeScript, you can set this to
                   * `ecmascript` instead, but there's no harm in leaving it
                   * as `typescript`.
                   *
                   * @see https://swc.rs/docs/configuration/compilation#jscparser
                   */
                  syntax: 'typescript',
                },
              },

              /**
               * The module configuration. This tells SWC how to output the
               * transpiled code.
               *
               * @see https://swc.rs/docs/configuration/modules
               */
              module: {
                /**
                 * This tells SWC to output CommonJS modules. MetaMask Snaps
                 * doesn't support ES modules yet, so this is necessary.
                 *
                 * @see https://swc.rs/docs/configuration/modules#commonjs
                 */
                type: 'commonjs',
              },
            },
          },
        },
      ],
    },

    /**
     * The resolve configuration. This tells Webpack how to resolve imports.
     * We set it to resolve `.js` and `.ts` files.
     *
     * @see https://webpack.js.org/configuration/resolve/
     */
    resolve: {
      /**
       * The extensions to resolve. We set it to resolve `.js` and `.ts`
       * files.
       */
      extensions: ['.js', '.ts'],
    },

    /**
     * The plugins to use.
     *
     * @see https://webpack.js.org/configuration/plugins/
     */
    plugins: [
      /**
       * The `SnapsWebpackPlugin` is a Webpack plugin that checks and updates
       * the manifest file, and evaluates the bundle in SES. While not strictly
       * required, it's highly recommended to use this plugin.
       */
      new SnapsWebpackPlugin({
        manifestPath: config.manifest.path,
        writeManifest: config.manifest.update,
        eval: options.evaluate ?? config.evaluate,
      }),

      /**
       * The `WatchPlugin` is a Webpack plugin that adds extra files to watch
       * for changes. This is useful for rebuilding the bundle when the
       * manifest file changes.
       */
      new WatchPlugin({
        files: [config.manifest.path],
      }),
    ],

    /**
     * The optimization configuration. This tells Webpack how to optimize the
     * bundle. Most of the time, you won't need to change this, as the default
     * options set by the `mode` option are sufficient.
     */
    optimization: {
      /**
       * The minimizer to use. We set it to use the `TerserPlugin`, which
       * minifies the bundle using SWC.
       */
      minimizer: [
        new TerserPlugin({
          minify: TerserPlugin.swcMinify,
        }),
      ],
    },

    /**
     * The infrastructure logging configuration. This tells Webpack how to
     * log messages.
     *
     * @see https://webpack.js.org/configuration/infrastructure-logging
     */
    infrastructureLogging: {
      /**
       * The level of logging to use. We set it to `none`, so that we can
       * control the output ourselves.
       */
      level: 'none',
    },
  };
}

/**
 * Get a Webpack compiler for the given config.
 *
 * @param config - The config object.
 * @param options - The Webpack options.
 * @returns The Webpack compiler.
 */
export function getCompiler(
  config: ProcessedWebpackConfig,
  options?: WebpackOptions,
) {
  const baseWebpackConfig = getDefaultConfiguration(config, options);
  const webpackConfig =
    config.customizeWebpackConfig?.(baseWebpackConfig) ?? baseWebpackConfig;

  const compiler = webpack(webpackConfig);
  compiler.hooks.afterDone.tap('AfterCompilePlugin', (stats) => {
    if (!stats) {
      return;
    }

    const { modules, time, errors } = stats.toJson();
    if (!modules || !time) {
      logError('Compilation status unknown. Please check your config.');
      process.exitCode = 1;
      return;
    }

    if (errors?.length) {
      logError(
        `Compiled ${modules?.length} files in ${time}ms with ${errors?.length} errors.`,
      );
      process.exitCode = 1;
      return;
    }

    logInfo(`Compiled ${modules?.length} files in ${time}ms.`);
  });

  return compiler;
}
