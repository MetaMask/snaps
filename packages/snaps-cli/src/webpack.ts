import SnapsWebpackPlugin from '@metamask/snaps-webpack-plugin';
import TerserPlugin from 'terser-webpack-plugin';
import { Configuration } from 'webpack';

import { ProcessedWebpackConfig } from './config';

/**
 * Get the Webpack devtool configuration based on the given snap config.
 *
 * - If `sourceMap` is `inline`, return `inline-source-map`.
 * - If `sourceMap` is `true`, return `source-map`.
 * - Otherwise, return `false`.
 *
 * @param config - The processed snap Webpack config.
 * @param config.sourceMap - Whether to generate source maps.
 * @returns The Webpack devtool configuration.
 */
function getDevTool({
  sourceMap,
}: ProcessedWebpackConfig): Configuration['devtool'] {
  if (sourceMap === 'inline') {
    return 'inline-source-map';
  }

  if (sourceMap === true) {
    return 'source-map';
  }

  return false;
}

/**
 * Get the default Webpack configuration. This is the configuration that will
 * be used if the user doesn't provide a custom Webpack configuration. The
 * configuration is based on the snap config.
 *
 * The default configuration includes:
 *
 * - `SWC` to transpile TypeScript and JavaScript files.
 * - `TerserPlugin` to minify the bundle.
 * - `SnapsWebpackPlugin` to validate the bundle and update the manifest.
 *
 * It can be customized through the `customizeWebpackConfig` function in the
 * snap config, but in most cases, you shouldn't need to do that.
 *
 * @param config - The processed snap Webpack config.
 * @returns The default Webpack configuration.
 */
export function getDefaultConfiguration(
  config: ProcessedWebpackConfig,
): Configuration {
  return {
    /**
     * `web` is the default target for Webpack, so we don't need to set it, but
     * we do so here for clarity. This ensures that Webpack will build a bundle
     * that can be loaded in a browser.
     *
     * @see https://webpack.js.org/configuration/target/
     */
    target: 'web',

    /**
     * The mode is set to `production` by default, so that Webpack will minify
     * and optimize the bundle.
     *
     * @see https://webpack.js.org/configuration/mode/
     */
    mode: 'production',

    /**
     * The entry point is set to the `entry` value from the config object.
     *
     * @see https://webpack.js.org/configuration/entry-context/
     */
    entry: config.entry,

    /**
     * The devtool option controls how source maps are generated. We set it to
     * the `sourceMap` value from the config object.
     *
     * @see https://webpack.js.org/configuration/devtool/
     */
    devtool: getDevTool(config),

    /**
     * The output options.
     *
     * @see https://webpack.js.org/configuration/output/
     */
    output: {
      /**
       * This indicates whether Webpack should clear the output directory
       * before building. We set it to the `clean` value from the config
       * object.
       *
       * @see https://webpack.js.org/configuration/output/#outputclean
       */
      clean: config.output.clean,

      /**
       * The filename of the bundle. We set it to the `filename` value from
       * the config object.
       *
       * @see https://webpack.js.org/configuration/output/#outputfilename
       */
      filename: config.output.filename,

      /**
       * The path to the output directory. We set it to the `path` value from
       * the config object.
       *
       * @see https://webpack.js.org/configuration/output/#outputpath
       */
      path: config.output.path,

      /**
       * The library configuration. This tells Webpack how to export the
       * bundle.
       *
       * @see https://webpack.js.org/configuration/output/#outputlibrary
       */
      library: {
        /**
         * This tells Webpack to export the bundle as a CommonJS module. This
         * is necessary for MetaMask Snaps.
         *
         * @see https://webpack.js.org/configuration/output/#outputlibrarytarget
         */
        type: 'commonjs',
      },
    },

    /**
     * The module configuration. This is where we tell Webpack how to handle
     * different types of files.
     *
     * @see https://webpack.js.org/configuration/module/
     */
    module: {
      rules: [
        {
          /**
           * We use the `swc-loader` to transpile TypeScript and JavaScript
           * files. This is a Webpack loader that uses the `SWC` compiler,
           * which is a much faster alternative to Babel and TypeScript's own
           * compiler.
           *
           * @see https://swc.rs/docs/usage/swc-loader
           */
          test: /\.tsx?$/u,
          exclude: /node_modules/u,
          use: {
            loader: 'swc-loader',

            /**
             * The options for the `swc-loader`. These can be overridden in the
             * `.swcrc` file.
             *
             * @see https://swc.rs/docs/configuration/swcrc
             */
            options: {
              parseMap: true,
              sourceMaps: true,
              inputSourceMap: true,
              /**
               * This tells SWC to generate source maps. We set it to the
               * `sourceMap` value from the config object.
               *
               * This must be enabled if source maps are enabled in the config.
               */
              // sourceMaps: Boolean(getDevTool(config)),

              jsc: {
                /**
                 * MetaMask targets ES2020, so we set the target to ES2020. If
                 * you need to support older browsers, you can set the target
                 * to something lower to include the necessary polyfills.
                 *
                 * @see https://swc.rs/docs/configuration/compilation#jsctarget
                 */
                target: 'es2020',

                parser: {
                  /**
                   * This tells the parser to parse TypeScript files. If you
                   * don't need to support TypeScript, you can set this to
                   * `ecmascript` instead, but there's no harm in leaving it
                   * as `typescript`.
                   *
                   * @see https://swc.rs/docs/configuration/compilation#jscparser
                   */
                  syntax: 'typescript',
                },
              },

              /**
               * The module configuration. This tells SWC how to output the
               * transpiled code.
               *
               * @see https://swc.rs/docs/configuration/modules
               */
              module: {
                /**
                 * This tells SWC to output CommonJS modules. MetaMask Snaps
                 * doesn't support ES modules yet, so this is necessary.
                 *
                 * @see https://swc.rs/docs/configuration/modules#commonjs
                 */
                type: 'commonjs',
              },
            },
          },
        },
      ],
    },

    /**
     * The resolve configuration. This tells Webpack how to resolve imports.
     * We set it to resolve `.js` and `.ts` files.
     *
     * @see https://webpack.js.org/configuration/resolve/
     */
    resolve: {
      /**
       * The extensions to resolve. We set it to resolve `.js` and `.ts`
       * files.
       */
      extensions: ['.js', '.ts'],
    },

    /**
     * The plugins to use.
     *
     * @see https://webpack.js.org/configuration/plugins/
     */
    plugins: [
      new SnapsWebpackPlugin({
        manifestPath: config.manifest.path,
        writeManifest: config.manifest.update,
      }),
    ],

    /**
     * The optimization configuration. This tells Webpack how to optimize the
     * bundle. Most of the time, you won't need to change this, as the default
     * options set by the `mode` option are sufficient.
     */
    optimization: {
      /**
       * The minimizer to use. We set it to use the `TerserPlugin`, which
       * minifies the bundle using SWC.
       */
      minimizer: [
        new TerserPlugin({
          minify: TerserPlugin.swcMinify,
        }),
      ],
    },
  };
}
