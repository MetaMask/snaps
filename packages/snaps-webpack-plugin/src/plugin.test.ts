// Allow Jest snapshots because the test outputs are illegible.
/* eslint-disable jest/no-restricted-matchers */

import webpack, { Stats, Configuration } from 'webpack';
import { createFsFromVolume, IFs, Volume } from 'memfs';
import { IPromisesAPI } from 'memfs/lib/promises';
import {
  DEFAULT_SNAP_BUNDLE,
  getSnapManifest,
} from '@metamask/snap-utils/test-utils';
import { checkManifest, evalBundle } from '@metamask/snap-utils';
import SnapsWebpackPlugin, { Options } from './plugin';

jest.mock('@metamask/snap-utils', () => ({
  ...jest.requireActual('@metamask/snap-utils'),
  evalBundle: jest.fn(),
  checkManifest: jest.fn(),
}));

type BundleOptions = {
  code?: string;
  options?: Options;
  fileSystem?: IFs;
  webpackOptions?: Configuration;
};

const bundle = async ({
  code = DEFAULT_SNAP_BUNDLE,
  options = { eval: false, manifestPath: undefined },
  fileSystem = createFsFromVolume(new Volume()),
  webpackOptions,
}: BundleOptions = {}): Promise<{
  code: string;
  fs: IPromisesAPI;
  stats: Stats;
}> => {
  const { promises: fs } = fileSystem;

  const bundler = webpack({
    mode: 'none',
    entry: {
      foo: '/foo.js',
    },
    output: {
      path: '/lib',
      filename: '[name].js',
    },
    plugins: [new SnapsWebpackPlugin(options)],
    ...webpackOptions,
  });

  bundler.inputFileSystem = fileSystem;
  bundler.outputFileSystem = fileSystem;

  await fs.mkdir('/lib', { recursive: true });
  await fs.writeFile('/foo.js', code);

  const outputStats = await new Promise<Stats>((resolve, reject) =>
    bundler.run((error, stats) => {
      if (error || !stats) {
        return reject(error);
      }

      return resolve(stats);
    }),
  );

  return {
    code: (await fs.readFile('/lib/foo.js', 'utf-8')) as string,
    fs,
    stats: outputStats,
  };
};

describe('SnapsWebpackPlugin', () => {
  it('processes files using Webpack', async () => {
    const { code } = await bundle();

    expect(code).toMatchSnapshot();
  });

  it('applies a transform', async () => {
    const { code } = await bundle({
      code: `
        // foo bar
        /* baz qux */
        const foo = 'bar';
      `,
    });

    expect(code).toMatchSnapshot();
    expect(code).not.toContain(`// foo bar`);
    expect(code).not.toContain(`/* baz qux */`);
  });

  it('forwards the options', async () => {
    const { code } = await bundle({
      code: `
        // foo bar
        /* baz qux */
        const foo = 'bar';
      `,
      options: {
        stripComments: false,
        eval: false,
        manifestPath: undefined,
      },
    });

    expect(code).toMatchSnapshot();
    expect(code).toContain(`// foo bar`);
    expect(code).toContain(`/* baz qux */`);
  });

  it('runs on the entire bundle', async () => {
    const fileSystem = createFsFromVolume(new Volume());
    const { promises: fs } = fileSystem;

    await fs.writeFile(
      '/bar.js',
      `
        // Returns baz
        export const bar = 'baz';
     `,
    );

    const { code } = await bundle({
      code: `
        import { bar } from './bar';

        // Sets foo to bar
        const foo = bar;
      `,
      fileSystem,
    });

    expect(code).toMatchSnapshot();
    expect(code).not.toContain(`// Sets foo to bar`);
    expect(code).not.toContain(`// Returns baz`);
  });

  it('generates a source map', async () => {
    const { fs } = await bundle({
      webpackOptions: {
        devtool: 'source-map',
      },
    });

    const map = await fs.readFile('/lib/foo.js.map', 'utf-8');
    expect(map).toMatchSnapshot();
  });

  it('evals the bundle if configured', async () => {
    const mock = evalBundle as jest.MockedFunction<typeof evalBundle>;
    mock.mockResolvedValue(null);

    await bundle({
      options: {
        eval: true,
        manifestPath: undefined,
      },
    });

    expect(mock).toHaveBeenCalledTimes(1);
    expect(mock).toHaveBeenCalledWith('/lib/foo.js');
  });

  it('checks the manifest if configured', async () => {
    const mock = checkManifest as jest.MockedFunction<typeof checkManifest>;
    mock.mockResolvedValue({
      manifest: getSnapManifest(),
      errors: [],
      warnings: [],
    });

    await bundle({
      options: {
        eval: false,
        manifestPath: '/snap.manifest.json',
      },
    });

    expect(mock).toHaveBeenCalledTimes(1);
    expect(mock).toHaveBeenCalledWith('/', true, expect.any(String));
  });

  it('does not fix the manifest if configured', async () => {
    const mock = checkManifest as jest.MockedFunction<typeof checkManifest>;
    mock.mockResolvedValue({
      manifest: getSnapManifest(),
      errors: [],
      warnings: [],
    });

    await bundle({
      options: {
        eval: false,
        manifestPath: '/snap.manifest.json',
        writeManifest: false,
      },
    });

    expect(mock).toHaveBeenCalledTimes(1);
    expect(mock).toHaveBeenCalledWith('/', false, expect.any(String));
  });

  it('logs manifest errors if writeManifest is disabled', async () => {
    const mock = checkManifest as jest.MockedFunction<typeof checkManifest>;
    mock.mockResolvedValue({
      manifest: getSnapManifest(),
      errors: ['foo', 'bar'],
      warnings: [],
    });

    await expect(
      bundle({
        options: {
          eval: false,
          manifestPath: '/snap.manifest.json',
          writeManifest: false,
        },
      }),
    ).rejects.toThrow('Manifest Error: The manifest is invalid.\nfoo\nbar');
  });

  it('logs manifest warnings', async () => {
    const mock = checkManifest as jest.MockedFunction<typeof checkManifest>;
    mock.mockResolvedValue({
      manifest: getSnapManifest(),
      errors: [],
      warnings: ['foo', 'bar'],
    });

    const { stats } = await bundle({
      options: {
        eval: false,
        manifestPath: '/snap.manifest.json',
        writeManifest: false,
      },
    });

    expect(stats.toJson().warnings?.[0].message).toMatch(
      'SnapsWebpackPlugin: Manifest Warning: Validation of snap.manifest.json completed with warnings.\n' +
        'foo\n' +
        'bar',
    );
  });

  it('forwards errors', async () => {
    const mock = evalBundle as jest.MockedFunction<typeof evalBundle>;
    mock.mockRejectedValue(new Error('foo'));

    await expect(
      bundle({
        options: {
          eval: true,
          manifestPath: undefined,
        },
      }),
    ).rejects.toThrow('foo');
  });
});
