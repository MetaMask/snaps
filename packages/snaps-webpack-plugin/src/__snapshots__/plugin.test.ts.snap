// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnapsWebpackPlugin applies a transform 1`] = `
"(() => {
  var __webpack_exports__ = {};
  const foo = 'bar';
})();"
`;

exports[`SnapsWebpackPlugin forwards the options 1`] = `
"/******/(() => {
  // webpackBootstrap
  var __webpack_exports__ = {};

  // foo bar
  /* baz qux */
  const foo = 'bar';

  /******/
})();"
`;

exports[`SnapsWebpackPlugin generates a source map 1`] = `"{"version":3,"file":"foo.js","names":["module","exports","onRpcRequest","request","console","log","method","id","__webpack_module_cache__","__webpack_require__","moduleId","cachedModule","undefined","__webpack_modules__","__webpack_exports__"],"sourceRoot":"","sources":["webpack://@metamask/snaps-webpack-plugin/../foo.js","webpack://@metamask/snaps-webpack-plugin/webpack/bootstrap","webpack://@metamask/snaps-webpack-plugin/webpack/before-startup","webpack://@metamask/snaps-webpack-plugin/webpack/startup","webpack://@metamask/snaps-webpack-plugin/webpack/after-startup"],"sourcesContent":["\\n  module.exports.onRpcRequest = ({ request }) => {\\n    console.log(\\"Hello, world!\\");\\n\\n    const { method, id } = request;\\n    return method + id;\\n  };\\n","// The module cache\\nvar __webpack_module_cache__ = {};\\n\\n// The require function\\nfunction __webpack_require__(moduleId) {\\n\\t// Check if module is in cache\\n\\tvar cachedModule = __webpack_module_cache__[moduleId];\\n\\tif (cachedModule !== undefined) {\\n\\t\\treturn cachedModule.exports;\\n\\t}\\n\\t// Create a new module (and put it into the cache)\\n\\tvar module = __webpack_module_cache__[moduleId] = {\\n\\t\\t// no module.id needed\\n\\t\\t// no module.loaded needed\\n\\t\\texports: {}\\n\\t};\\n\\n\\t// Execute the module function\\n\\t__webpack_modules__[moduleId](module, module.exports, __webpack_require__);\\n\\n\\t// Return the exports of the module\\n\\treturn module.exports;\\n}\\n\\n","","// startup\\n// Load entry module and return exports\\n// This entry module used 'module' so it can't be inlined\\nvar __webpack_exports__ = __webpack_require__(0);\\n",""],"mappings":";;IACEA,MAAA,CAAAC,OAAA,CAAAC,YAA2B;MAAMC;IAAA,CAAS;MAC5CC,OAAA,CAAAC,GAAA;MAEA;QAAYC,MAAA;QAAAC;MAAA,IAAaJ,OAAA;MACzB,OAAAG,MAAA,GAAAC,EAAA;IACA;;ECLA,IAAAC,wBAAA;EAGA,SAAAC,oBAAAC,QAAA;IAEA,IAAAC,YAAA,GAAAH,wBAAA,CAAAE,QAAA;IACA,IAAAC,YAAA,KAAAC,SAAA;MACA,OAAAD,YAAA,CAAAV,OAAA;IACA;IAEA,IAAAD,MAAA,GAAAQ,wBAAA,CAAAE,QAAA;MAGAT,OAAA;IACA;IAGAY,mBAAA,CAAAH,QAAA,EAAAV,MAAA,EAAAA,MAAA,CAAAC,OAAA,EAAAQ,mBAAA;IAGA,OAAAT,MAAA,CAAAC,OAAA;EACA;EEnBA,IAAAa,mBAAA,GAAAL,mBAAA"}"`;

exports[`SnapsWebpackPlugin processes files using Webpack 1`] = `
"(() => {
  var __webpack_modules__ = [module => {
    module.exports.onRpcRequest = ({
      request
    }) => {
      console.log("Hello, world!");
      const {
        method,
        id
      } = request;
      return method + id;
    };
  }];
  var __webpack_module_cache__ = {};
  function __webpack_require__(moduleId) {
    var cachedModule = __webpack_module_cache__[moduleId];
    if (cachedModule !== undefined) {
      return cachedModule.exports;
    }
    var module = __webpack_module_cache__[moduleId] = {
      exports: {}
    };
    __webpack_modules__[moduleId](module, module.exports, __webpack_require__);
    return module.exports;
  }
  var __webpack_exports__ = __webpack_require__(0);
})();"
`;

exports[`SnapsWebpackPlugin runs on the entire bundle 1`] = `
"(() => {
  "use strict";

  var __webpack_modules__ = [, (__unused_webpack_module, __webpack_exports__, __webpack_require__) => {
    __webpack_require__.r(__webpack_exports__);
    __webpack_require__.d(__webpack_exports__, {
      bar: () => bar
    });
    const bar = 'baz';
  }];
  var __webpack_module_cache__ = {};
  function __webpack_require__(moduleId) {
    var cachedModule = __webpack_module_cache__[moduleId];
    if (cachedModule !== undefined) {
      return cachedModule.exports;
    }
    var module = __webpack_module_cache__[moduleId] = {
      exports: {}
    };
    __webpack_modules__[moduleId](module, module.exports, __webpack_require__);
    return module.exports;
  }
  (() => {
    __webpack_require__.d = (exports, definition) => {
      for (var key in definition) {
        if (__webpack_require__.o(definition, key) && !__webpack_require__.o(exports, key)) {
          Object.defineProperty(exports, key, {
            enumerable: true,
            get: definition[key]
          });
        }
      }
    };
  })();
  (() => {
    __webpack_require__.o = (obj, prop) => Object.prototype.hasOwnProperty.call(obj, prop);
  })();
  (() => {
    __webpack_require__.r = exports => {
      if (typeof Symbol !== 'undefined' && Symbol.toStringTag) {
        Object.defineProperty(exports, Symbol.toStringTag, {
          value: 'Module'
        });
      }
      Object.defineProperty(exports, '__esModule', {
        value: true
      });
    };
  })();
  var __webpack_exports__ = {};
  (() => {
    __webpack_require__.r(__webpack_exports__);
    var _bar__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(1);
    const foo = _bar__WEBPACK_IMPORTED_MODULE_0__.bar;
  })();
})();"
`;
